rllib/core/optim/tests/test_rl_optimizer.py

```diff
@@ -0,0 +1,386 @@
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        results = {
+            "avg_reward": batch["rewards"].mean(),
+        }
+        results.update(loss_numpy)
+        results["module_norm"] = model_norm(self._module)
+        return results
+
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        torch_batch = convert_to_torch_tensor(batch)
+        fwd_out = self._module.forward_train(torch_batch)
+        loss = self._rl_optimizer.compute_loss(torch_batch, fwd_out)
+        gradients = self.compute_gradients(loss)
+        post_processed_gradients = self.on_after_compute_gradients(gradients)
+        self.apply_gradients(post_processed_gradients)
+        results = self.compile_results(batch, fwd_out, loss, post_processed_gradients)
+        return results
+
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]]
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module
+
+            For example compute and apply gradients to self._module if
+                necessary.
+
+        Args:
+            loss: variable(s) used for optimizing self._module.
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+        # for torch
+        if isinstance(loss, dict):
+            loss = loss["total_loss"]
+        loss.backward()
+        grads = {n: p.grad for n, p in self._module.named_parameters()}
+        return grads
+
+    def apply_gradients(self, gradients: Mapping[str, Any]) -> None:
+        """Perform an update on self._module"""
+        for optimizer in self._rl_optimizer.get_optimizers().values():
+            optimizer.step()
+            optimizer.zero_grad()
+
+    def set_state(self, state: Mapping[str, Any]) -> None:
+        """Set the state of the trainer."""
+        self._rl_optimizer.set_state(state.get("optimizer_state", {}))
+        self._module.set_state(state.get("module_state", {}))
+
+    def get_state(self) -> Mapping[str, Any]:
+        """Get the state of the trainer."""
+        return {
+            "module_state": self._module.get_state(),
+            "optimizer_state": self._rl_optimizer.get_state(),
+        }
+
+
+class TestRLOptimizer(unittest.TestCase):
+    def test_rl_optimizer_example_torch(self):
+        ray.init()
```

Comment:
please ``ray.init()`` in ``def setUpClass(cls)`` and ``ray.shutdown()`` in ``def tearDownClass(cls)``
https://github.com/ray-project/ray/blob/master/rllib/algorithms/appo/tests/test_appo.py#L16-L22

this is assuming test_rl_optimizer_example_torch will always run before the next test, which may not be true.